File: May_26th/Task_4.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

namespace Threads
{
    class ThreadwithDynamic
    {
        private static Queue<string> workQueue = new Queue<string>();
        private static readonly object lockObj = new object();
        private static bool done = false;
        public static void WorkerThread()
        {
            while (true)
            {
                string workItem = null;
                lock (lockObj)
                {
                    if (workQueue.Count > 0)
                    {
                        workItem = workQueue.Dequeue();
                    }
                    else if (done)
                    {
                        break;
                    }
                }
                if (workItem != null)
                {
                    Console.WriteLine($"Processing {workItem}");
                    Thread.Sleep(1000); // Simulate work
                }
                else
                {
                    Thread.Sleep(100); // Avoid busy waiting
                }
            }
            Console.WriteLine("Worker finished all tasks.");
        }
        public static void Main()
        {
            Thread worker = new Thread(WorkerThread);
            worker.Start();
            for (int i = 0; i < 5; i++)
            {
                lock (lockObj)
                {
                    workQueue.Enqueue($"Task {i}");
                }

                Thread.Sleep(300);
            }
            lock (lockObj)
            {
                done = true;
            }
            worker.Join();

        }
    }
}

File: May_31th/Task-4.c
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ConsoleApp7
{
    class Task_4
    {

        static int totalLines = 0;
        static object lockObj = new object();

        static void CountLines(string filePath)
        {
            try
            {

                int count = File.ReadAllLines(filePath).Length;


                lock (lockObj)
                {
                    totalLines += count;
                }

                Console.WriteLine($"{filePath}: {count} lines");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error reading {filePath}: {ex.Message}");
            }
        }

        static void Main()
        {

            string[] files = { "file1.txt", "file2.txt", "file3.txt" };

            Thread[] threads = new Thread[files.Length];

            for (int i = 0; i < files.Length; i++)
            {
                string path = files[i]; 
                threads[i] = new Thread(() => CountLines(path));
                threads[i].Start();
            }


            foreach (var t in threads)
            {
                t.Join();
            }

            Console.WriteLine($"\nTotal lines across all files: {totalLines}");
        }
    }
}

